### PR TITLE
raft: leader respond to learner read index message

### PR DESCRIPTION
**Pull request by nolouch**
In a cluster with 1 leader and 1 learner, read request in learner will get an empty read index.
pick from: https://github.com/etcd-io/etcd/pull/10590

### CODE REVIEW

**Comment by siddontang** [comment]
Great Catch, CI failed

File: src/raft.rs [hunk 1]
```diff
@@ -1691,12 +1691,19 @@ impl<T: Storage> Raft<T> {
...
                         index: self.raft_log.committed,
                         request_ctx: m.take_entries()[0].take_data(),
                     };
                     self.read_states.push(rs);
+                } else {
+                    let mut to_send = Message::default();
+                    to_send.set_to(m.get_from());
+                    to_send.set_msg_type(MessageType::MsgReadIndexResp);
+                    to_send.set_index(self.raft_log.committed);
+                    to_send.set_entries(m.take_entries());
```
Comment (ice1000) [inline src/raft.rs]: Try:

```rust
to_send.to = m.from;
to_send.set_msg_type(MessageType::MsgReadIndexResp);
to_send.index = self.raft_log.committed;
to_send.entries = m.take_entries();
```

File: tests/integration_cases/test_raft.rs [hunk 2]
```diff
@@ -2189,6 +2189,75 @@ fn test_read_only_option_safe() {
...
+        let rs = &read_states[0];
+        if rs.index != wri {
+            panic!("#{}: read_index = {}, want {}", i, rs.index, wri)
+        }
+        let vec_wctx = wctx.as_bytes().to_vec();
+        if rs.request_ctx != vec_wctx {
+            panic!(
+                "#{}: request_ctx = {:?}, want {:?}",
+                i, rs.request_ctx, vec_wctx
+            )
```
Comment (ice1000) [inline tests/integration_cases/test_raft.rs]: Why not:

```rust
assert_ne!(rs.request_ctx, vec_wctx,
  "#{}: request_ctx = {:?}, want {:?}",
  i, rs.request_ctx, vec_wctx);
```

File: tests/integration_cases/test_raft.rs [hunk 3]
```diff
@@ -2189,6 +2189,75 @@ fn test_read_only_option_safe() {
...
+            .unwrap()
+            .read_states
+            .drain(..)
+            .collect();
+        if read_states.is_empty() {
+            panic!("#{}: read_states is empty, want non-empty", i);
+        }
+        let rs = &read_states[0];
+        if rs.index != wri {
+            panic!("#{}: read_index = {}, want {}", i, rs.index, wri)
```
Comment (ice1000) [inline tests/integration_cases/test_raft.rs]: Why not:

```rust
assert_ne!(rs.index, wri,
  "#{}: read_index = {}, want {}",
  i, rs.index, wri);
```

File: src/raft.rs [hunk 4]
```diff
@@ -1691,12 +1691,19 @@ impl<T: Storage> Raft<T> {
                             }
                         }
                     }
-                } else {
+                } else if m.get_from() == INVALID_ID || m.get_from() == self.id {
```
Comment (Hoverbear) [inline src/raft.rs]: Could you please include the comments like in the etcd code? This way we can know why we did this later.

File: src/raft.rs (hunk 1 above)
Comment (hicqu) [reply src/raft.rs]: For the style, I agree with @ice1000 , it looks better.
But I think we should unify the style later, after https://github.com/tikv/protobuf-build/pull/7 is merged and some discussion about those options in `raft-rs`.

Comment (ice1000) [reply src/raft.rs]: That's reasonable too. This comment is more like a "FYI you can do this" comment.

Comment (nolouch) [reply src/raft.rs]: ok, let's do it in another PR.

File: src/raft.rs (hunk 4 above)
Comment (nolouch) [reply src/raft.rs]: done.

**Comment by nolouch** [comment]
PTAL  @Hoverbear @hicqu

**Comment by hicqu** [comment]
LGTM

**Comment by hicqu** [comment]
LGTM

**Pull request merged**